### PR TITLE
Detail page commit list diffstat

### PR DESCRIPTION
**Pull request by Eclipse-Dominator**
### Summary:
Partially addresses #69, depends on #132 

### Changes Made:
- Adds diffstats component to display line additions and deletions
- Adds RSQ data structures to enable fast RSQ queries
- Calculates cumulative stats in the commits
- Adds the ability to check cumulative stats across multiple commits
- Show currently shown commits

### Commit Message:

```
The commit list currently doesn't display line changes
Let's display the line change per commit as well as the cumulative line changes

- Adds diffstats component to display line additions and deletions
- Adds RSQ data structures to enable fast RSQ queries
- Calculates cumulative stats in the commits
- Adds the ability to check cumulative stats across multiple commits
- Show currently shown commits
```

### CODE REVIEW

File: src/app/user-detail-viewer/commit-list/commit-list.component.ts [hunk 1]
```diff
@@ -13,6 +14,29 @@ export interface DateRange {
...
+    this.deletions += o.deletions;
+    this.count += o.count;
+  }
+
+  sub(o: CumulativeStats): void {
+    this.additions -= o.additions;
+    this.deletions -= o.deletions;
+    this.count -= o.count;
+  }
+}
```
Comment (gycgabriel) [inline src/app/user-detail-viewer/commit-list/commit-list.component.ts:39]: Create a model.ts for this

File: src/app/user-detail-viewer/commit-list/commit-list.component.ts [hunk 2]
```diff
@@ -56,13 +82,23 @@ export class CommitListComponent implements OnInit {
...
 
     const commitToIndex = (commit: GithubCommit) => this.dateToIndex(commit.committedDate);
 
     this.endIndex = commitToIndex(this.commitList[this.commitList.length - 1]);
+    const prefixArr = new Array(this.endIndex + 1);
+    for (let i = 0; i < prefixArr.length; i++) {
+      prefixArr[i] = new CumulativeStats();
+    }
+    for (const commit of this.commitList) {
+      prefixArr[commitToIndex(commit)].add(new CumulativeStats(commit));
```
Comment (gycgabriel) [inline src/app/user-detail-viewer/commit-list/commit-list.component.ts:98]: I checked your `dateToIndex` function, it uses a sparse array. Tested on cs2103, it works fine without any jitters. We can go with this. 👍

**Review by gycgabriel — CHANGES_REQUESTED** [review]

**Review by gycgabriel — COMMENTED** [review]
Tested and works well! Some comments to look into.

Found a bug when switching repo on detail page, will need to disable that button.

File: src/app/user-detail-viewer/commit-list/commit-list.component.ts (hunk 2 above)
Comment (gycgabriel) [reply src/app/user-detail-viewer/commit-list/commit-list.component.ts:98]: For reference, the average size of the array will be from 40-50

File: src/app/core/models/datastructure/rsq.model.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  get(index: number): T {
+    return this.rsq(index, index);
+  }
+}
+
+/**
+ * 1-Indexed Prefix Sum to abstract away rsq operations
+ */
+export class PrefixSum<T extends Accumulator<T>> extends RangeQueries<T> {
+  constructor(private arr: T[], private identity: () => T) {
```
[the remark this answers is not in the document]
Comment (Eclipse-Dominator) [reply src/app/core/models/datastructure/rsq.model.ts:42]: having an identity function here is mainly cos the accumulator interface suggests that the object is mutable. To perform a range query, we either need to create an "identity" accumulator or have way to clone the accumulator.

**Comment by Eclipse-Dominator** [comment]
> Tested and works well! Some comments to look into.
> 
> Found a bug when switching repo on detail page, will need to disable that button in the future

This issue appeared after I switched to using phases for page swap. I think I will make a separate PR in the Profile Page branch to address this before merging it.

File: src/app/core/models/datastructure/rsq.model.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    super();
+    for (let i = 1; i < arr.length; i++) {
+      arr[i].add(arr[i - 1]);
```
[the remark this answers is not in the document]
Comment (Eclipse-Dominator) [reply src/app/core/models/datastructure/rsq.model.ts:45]: I made the constructor take in an array that is to be used as the prefix sum so I think I will mention it in the documentation.

**Review by gycgabriel — APPROVED** [review]
LGTM!

**Pull request merged**